File: AlternativeQL/src/org/ql/ast/expression/literal/DecimalLiteral.java
package org.ql.ast.expression.literal;

import org.ql.ast.Expression;
import org.ql.ast.expression.ExpressionVisitor;

import java.math.BigDecimal;

public class DecimalLiteral extends Expression {
    private final BigDecimal value;

    public DecimalLiteral(BigDecimal value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return value.toPlainString();
    }

    @Override
    public <T, C> T accept(ExpressionVisitor<T, C> visitor, C context) {
        return visitor.visitDecimal(this, context);
    }

    public BigDecimal getValue() {
        return value;
    }
}

File: OffByOne/src/OffByOne.Qls/Ast/Style/Literals/DecimalLiteral.cs
namespace OffByOne.Qls.Ast.Style.Literals
{
    using System;

    using MoreDotNet.Wrappers;

    using OffByOne.Qls.Ast.Style.Literals.Base;
    using OffByOne.Qls.Common.Visitors.Contracts;

    public class DecimalLiteral : Literal
    {
        public DecimalLiteral(double value)
        {
            this.Value = value;
        }

        public DecimalLiteral(string value)
        {
            if (value.IsNullOrWhiteSpace())
            {
                throw new ArgumentException("A non-null, non-empty string must be given", nameof(value));
            }

            this.Value = double.Parse(value);
        }

        public double Value { get; }

        public override TResult Accept<TResult, TContext>(
            ILiteralVisitor<TResult, TContext> visitor,
            TContext context)
        {
            return visitor.Visit(this, context);
        }
    }
}
